### PR TITLE
Modernize infrastructure

### LINKED ISSUE #42: migrate docs to GH Pages?
The [RTD](https://spglm.readthedocs.io/en/latest/?badge=latest) build is way out of date. As part of the modernization effort, should we migrate to GitHub Pages?

cc @weikang9009 @martinfleis @ljwolf @knaaptime

### LINKED ISSUE #43: Modernize infrastructure
Modernize infrastructure à la https://github.com/pysal/giddy/issues/200, etc.

### LINKED ISSUE #44: add a `codecov` config file
Add a `codecov` config file. See example in `giddy` --> [`codecov.yml`](https://github.com/pysal/giddy/blob/main/codecov.yml)

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR:
* resolves #42
* resolves #43
* resolves #44 
  * swaps out `setup.*` for `pyproject.toml`
  * adopt `setuptools_scm`
  * `ruff` lints more strictly
  * adds `pre-commit-config`
  * cleans and reruns notebooks
  * remove old/superfluous files
    * all `setup*.*` (and related) files
    * all `docs/{build,generated}*` files (*this accounts for the huge file diff*)
  * adopts `pytest` over `unittest`
  * revamps docs & ensure docs build locally (and beautifully)
  * etc.

@martinfleis this isn't complete, but how's looking so far? Some questions/comments:
* Do we want to address #42 in this PR and add an action for it? And if so, shall we remove the RTD yaml?
* Should I put a release action in here even though we won't be able to use it yet?
* Anything obvious I am missing here?

### CODE REVIEW

File: spglm/base.py
Comment (martinfleis) [inline spglm/base.py:23]: Tricky change as until now, passing additional kwargs did not raise but now it will. I would revert it.

**Review by martinfleis — COMMENTED** [review]
Great job!

File: spglm/base.py
Comment (jGaboardi) [reply spglm/base.py:23]: Done.

**Comment by jGaboardi** [comment]
@martinfleis I think this PR is ready for review now. I have added the docs and release workflows, event though the release will not work until we get access to [`PyPI/spglm`](https://pypi.org/project/spglm/).

**Comment by martinfleis** [comment]
`readthedocs.yml` can go, so we don't build the docs twice, now that we have the action. Otherwise looks good, though it is a bit pain to review this :D.

**Comment by jGaboardi** [comment]
> `readthedocs.yml` can go, so we don't build the docs twice, now that we have the action. Otherwise looks good, though it is a bit pain to review this :D.

No good deed goes unpunished. 🤣

**Comment by jGaboardi** [comment]
@martinfleis And it's been a long time since I've set up the GH-Pages stuff. Not sure I did it correctly... In [Setting/Pages](https://github.com/pysal/spglm/settings/pages) I simply changed the branch to `main/root`, but that doesn't feel right...

**Comment by martinfleis** [comment]
I've never done it (momepy still runs on RTD and its own domain) but you'll need to point it to `gh-pages/root`. But `gh-pages` branch does not exist yet. Not sure if the action will create it or it needs to be created and populated manually first.

**Comment by jGaboardi** [comment]
Perhaps @knaaptime can advise? But also, once this is reviewed and merged I can simply do some trial and error `v1.1.0rcXX` tag & releases (which will fail) to get the doc build sorted.

**Comment by knaaptime** [comment]
pretty sure the action will create the branch if necessary

**Comment by jGaboardi** [comment]
> @jGaboardi it seems that you should have all the rights you need to setup the publishing action with a token, right?

Yep, I just set that up right now. So once this PR is reviewed and merged we can give `v1.1.0rc1` a go.

**Comment by martinfleis** [comment]
@jGaboardi What is the status here?

**Comment by jGaboardi** [comment]
Waiting for a review.

**Review by martinfleis — APPROVED** [review]

**Pull request merged**
